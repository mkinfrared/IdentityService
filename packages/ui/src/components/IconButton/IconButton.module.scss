@import "../../styles/base/baseButton";

.IconButton {
  @extend %buttonBase;

  align-items: center;
  aspect-ratio: 1/1;
  border-radius: 50%;
  display: inline-flex;
  justify-content: center;
  min-width: auto;
  position: relative;

  &.loading {

    .spinner {
      opacity: 1;
    }
  }

  svg {
    fill: var(--color);
    height: var(--text-size);
  }

  .spinner {
    height: 100%;
    inset: 0;
    opacity: 0;
    position: absolute;
  }
}

.Story {
}
