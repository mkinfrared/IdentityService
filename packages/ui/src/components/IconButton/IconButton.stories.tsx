import { Meta, Story } from "@storybook/react";
import React from "react";

import Button from "components/Button";
import { ReactComponent as Clear } from "icons/clear.svg";
import { ReactComponent as Doc } from "icons/description.svg";
import { ReactComponent as Eye } from "icons/visibility.svg";

import { IconButton } from "./IconButton";
import css from "./IconButton.module.scss";
import { IconButtonProps } from "./IconButton.type";

export default {
  title: "UI/IconButton",
  component: IconButton,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const Template: Story<IconButtonProps> = (args) => (
  <div className={css.Story}>
    <IconButton {...args}>
      <Eye />
    </IconButton>
    <IconButton {...args}>
      <Doc />
    </IconButton>
    <IconButton {...args}>
      <Clear />
    </IconButton>
    <Button>Button</Button>
  </div>
);

const Default = Template.bind({});

Default.args = {
  disabled: false,
  variant: "opaque",
};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
