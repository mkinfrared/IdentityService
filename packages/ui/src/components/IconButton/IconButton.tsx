/* eslint-disable react/button-has-type */
import { classNames } from "@identity-service/core";
import React, { memo } from "react";

import Spinner from "components/Spinner";

import css from "./IconButton.module.scss";
import { IconButtonProps } from "./IconButton.type";

const IconButton = ({
  children,
  className,
  disabled = false,
  loading = false,
  type = "button",
  variant = "opaque",
  ...rest
}: IconButtonProps) => {
  const classes = [css.Button, className];

  switch (variant) {
    case "transparent":
      classes.push(css.transparent);

      break;

    default:
      classes.push(css.opaque);
  }

  if (loading) {
    classes.push(css.loading);
  }

  return (
    <button
      className={classNames(css.IconButton, ...classes)}
      disabled={disabled}
      data-testid="IconButton"
      type={type}
      {...rest}
    >
      <Spinner className={css.spinner} type="oval" />
      {children}
    </button>
  );
};

export { IconButton };

export default memo(IconButton);
