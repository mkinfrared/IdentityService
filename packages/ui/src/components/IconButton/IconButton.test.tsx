import { render } from "@testing-library/react";
import React from "react";

import { IconButton } from "./IconButton";

describe("<IconButton />", () => {
  const Component = <IconButton />;

  it("should be defined", () => {
    expect(IconButton).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("IconButton");

    expect(element).toBeDefined();
  });

  it("should render an opaque button by default", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("IconButton");

    expect(element).toBeDefined();

    expect(element).toHaveClass("opaque");
  });

  it("should render a transparent button when variant prop is transparent", () => {
    const { getByTestId } = render(<IconButton variant="transparent" />);
    const element = getByTestId("IconButton");

    expect(element).toBeDefined();

    expect(element).toHaveClass("transparent");
  });

  it("should render a disabled button when disabled prop is true", () => {
    const { getByTestId } = render(<IconButton disabled />);
    const element = getByTestId("IconButton");

    expect(element).toBeDefined();

    expect(element).toBeDisabled();
  });

  it("should add class name passed in props", () => {
    const classname = "marklar";
    const { getByTestId } = render(<IconButton className={classname} />);
    const element = getByTestId("IconButton");

    expect(element).toBeDefined();

    expect(element).toHaveClass(classname);
  });

  it("should add 'loading' class name to button", () => {
    const classname = "loading";

    const { getByTestId } = render(
      <IconButton className={classname} loading />,
    );

    const element = getByTestId("IconButton");

    expect(element).toBeDefined();

    expect(element).toHaveClass(classname);
  });
});
