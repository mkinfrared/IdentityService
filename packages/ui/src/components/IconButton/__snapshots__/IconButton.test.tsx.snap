// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> should match the snapshot 1`] = `
<div>
  <button
    class="IconButton Button opaque"
    data-testid="IconButton"
    type="button"
  >
    <svg
      class="Spinner spinner"
    >
      oval.svg
    </svg>
  </button>
</div>
`;
